umbral/umbral.py

```diff
@@ -120,98 +119,124 @@ def __bytes__(self):
+                Passing both is also fine.")
+
+        self._point_eph_e = point_eph_e
+        self._point_eph_v = point_eph_v
+        self._bn_sig = bn_sig
+
+        self._point_eph_e_prime = e_prime
+        self._point_eph_v_prime = v_prime
+        self._point_noninteractive = noninteractive_point
 
         self.cfrags = {}
+        self._contents = None
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    class NotValid(ValueError):
+        """
+        raised if the capusle does not pass verification.
+        """
+
+    @classmethod
+    def from_bytes(cls, data: bytes, curve: ec.EllipticCurve, is_reconstructed=False):
         """
         Instantiates a Capsule object from the serialized data.
         """
-        eph_e = Point.from_bytes(data[0:33], curve)
-        eph_v = Point.from_bytes(data[33:66], curve)
-        sig = BigNum.from_bytes(data[66:98], curve)
+        if is_reconstructed:
+            e_prime = Point.from_bytes(data[0:33], curve)
+            v_prime = Point.from_bytes(data[33:66], curve)
+            eph_ni = Point.from_bytes(data[66:99], curve)
 
-        return Capsule(eph_e, eph_v, sig)
+            return cls(e_prime=e_prime, v_prime=v_prime, noninteractive_point=eph_ni)
+        else:
+            eph_e = Point.from_bytes(data[0:33], curve)
+            eph_v = Point.from_bytes(data[33:66], curve)
+            sig = BigNum.from_bytes(data[66:98], curve)
 
-    def to_bytes(self):
+            return cls(eph_e, eph_v, sig)
+
+    @property
+    def contents(self):
+        return self._contents
+
+    def to_bytes(self, reconstructed_components=False):
         """
         Serialize the Capsule into a bytestring.
         """
-        eph_e = self.point_eph_e.to_bytes()
-        eph_v = self.point_eph_v.to_bytes()
-        sig = self.bn_sig.to_bytes()
+        if reconstructed_components:
+            eph_e = self._point_eph_e_prime
+            eph_v = self._point_eph_v_prime
+            point_noninter = self._point_noninteractive
+
+            return eph_e.to_bytes() + eph_v.to_bytes() + point_noninter.to_bytes()
+        else:
+            eph_e = self._point_eph_e.to_bytes()
+            eph_v = self._point_eph_v.to_bytes()
+            sig = self._bn_sig.to_bytes()
 
-        return eph_e + eph_v + sig
+            return eph_e + eph_v + sig
 
     def verify(self, params: UmbralParameters):
 
-        e = self.point_eph_e
-        v = self.point_eph_v
-        s = self.bn_sig
+        e = self._point_eph_e
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return s * params.g == v + (h * e)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def open(self, pub_bob, priv_bob, pub_alice, force_reopen=False, pre=None):
```

Comment:
Are we sure we want to provide such a method? I like the name (`open`), but not sure if I like what's happening here.

I do prefer using the simple API to do these sorts of things now and would prefer to keep functions that handle these kinds of things there.

Reply:
Yeah, it's a good question.  Ideally, the "simple API" calls the one-and-only public read method on this class, right?  I was trying to make `open` that function.  Then additionally, I also like the idea of caching the result so that the API doesn't have to concern itself with that.  

So the question is: is this the one obvious way to do this?  Or is there a better one?

Reply:
Well, if we provide an `open` does that mean there should be a `close`?

Because we can't reliably provide a closing method to clear memory and handle all the sensitive components in Python. My thinking is that we should just urge the user to use the API as provided. If we see evidence of users doing otherwise in the future, maybe make an adjustment then?

Reply:
So you are thinking of removing this and just promoting `_decapsulate_reencryoted` to public?

I'm not entirely opposed to that, but there are a couple of ways in which I think this approach is superior:

1) it allows the user (or perhaps us, in a future API revision) to separate the decapsulation from accessing the contents.  This seems like kind of a big deal to me, because they may want to fail fast on the decapsulation, but not access the contents until the end of the loop or whatever.
2) it keeps the mechanics of decapsulation in a private layer, while putting the access in a public function. I think this is just good engineering generally.

As for the "close" thing: that occurred to me too, but I actually think that it serves to obfuscate, rather than clarify, the metaphor.  Once you've dug up a time capsule and removed the contents, do you expect the capsule to provide a facility for closing it again?

Reply:
I actually really like the idea of using `decapsulate_reencrypted` as a public method.

As for your first point, decapsulation is the access of the contents. We should not confuse this to the user who might think, "Oh, I haven't decrypted it, I have just decapsulated it."

For your second, it is a good point, but I don't think hiding decapsulating from the user is going to do us any good. Users are already comfortable with decryption, decapsulating is pretty similar and I don't think is too hard to grasp the concept of.

Again, I really want to stress to users that decapsulating _is_ decryption and they should use this process similarly. Accessing the key material is a big deal like that.

Reply:
OK, this is all making sense to me.

Here's what is still weird:

* Why do we have *both* `PRE._decapsulate_reencrypted` *and* `PRE.decapsulate_reencrypted` ?

I think it makes sense to:

* Move the latter to `Capsule`
* Rename it to something more straightforward, like, `get_contents`
* Rename `PRE._decapsulate_reencrypted` to `PRE.decapsulate_reencrypted` 

Then we have a clean, simple API in `PRE`, and a good clean metaphor for `Capsule`.

How's that sound?

Reply:
If you want to see what that looks like, I just pushed it in f3be38b63446cd771fecb2c74285ecbc6b1acb56.  For my money, I think it's getting pretty close to being a go.